tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc

```diff
@@ -131,13 +146,29 @@ class LowerAddNOp : public OpRewritePattern<TF::AddNOp> {
     // support variant type so variant types require special handling.
     if (getElementTypeOrSelf(op.getType()).isa<VariantType>()) return failure();
 
-    // TODO(hinsu): Improve parallelism by splitting operands in two halves and
-    // accumulating them first.
-    Value result = *op.inputs().begin();
-    for (Value operand : llvm::drop_begin(op.inputs(), 1)) {
-      result = rewriter.create<TF::AddV2Op>(op.getLoc(), result, operand);
+    auto begin = op.inputs().begin();
+    // Return the only operand directly.
+    if (op.N() == 1) {
+      rewriter.replaceOp(op, *begin);
+      return success();
     }
 
+    // Helper functor to accumulate from `begin` to `end` (exclusive).
+    auto accumulate_add = [&rewriter, &op] (auto begin, auto end) -> Value {
+      Value result = *begin;
+      ++begin;
+      for (auto operand = begin; operand != end; ++operand) {
+        result = rewriter.create<TF::AddV2Op>(op.getLoc(), result, *operand);
+      }
+      return result;
+    };
+
+    // Accumulate range `[begin, half)` and `[half, end)`,
+    // and add the results of two halves.
```

Comment:
Why two halves vs a tree of additions? (meaning this is an improvement but if the goal is additional parallelism one could go further)

Reply:
Sounds good! I change it to tree-based reduction. Please check it out 😃